### PR TITLE
CNS-368: fixation store fix refcount and deletions

### PR DESCRIPTION
**Pull request by orenl-lava**
Fix refcount handling: FixationStore entry refcount was half-baked: the latest entries had refcount 0, and the code had special handling to avoid ignoring/removing them.

Fix garbage collection logic for stale entries: (1) When entries get StaleAt field set, set a timer to fire on that block; (2) Whenever such timer fires, scan all versions of the respective index to remove eligible stale versions ("eligible" stale version is one whose preceding version is stale; the oldest version is always "eligible").

### CODE REVIEW

File: common/fixation_entry.go [hunk 1]
```diff
@@ -169,42 +166,66 @@ func (fs *FixationStore) AppendEntry(
...
 	iterator := sdk.KVStorePrefixIterator(store, []byte{})
 	defer iterator.Close()
 
-	for iterator.Valid() {
-		// umarshal the old entry version
-		var oldEntry types.Entry
-		fs.cdc.MustUnmarshal(iterator.Value(), &oldEntry)
+	// remove "eligible" stale entry versions, where "eligible" is either:
+	//   an entry version that follows a stale entry version, -OR-
+	//   the older entry version
```
Comment (oren-lava) [inline common/fixation_entry.go:185]: oldest?

File: common/fixation_entry.go [hunk 2]
```diff
@@ -169,42 +166,66 @@ func (fs *FixationStore) AppendEntry(
...
+	// rationale: we keep certain stale entry versions in order to mark -
+	// for future findEntry - that blocks from that stale entry onward are
+	// stale (otherwise, a lookup might resolve successfully with an older
+	// non-stale entry version). For this, one - the oldest - marker is
```
Comment (oren-lava) [inline common/fixation_entry.go:189]: why would we need the markers to mark which entry is stale? the refCount + STALE_TIME mechanism does just that. Also, what if we want to keep the oldest version of the entry? In other words, why is the oldest entry automatically stale?

File: common/fixation_entry.go [hunk 3]
```diff
@@ -256,7 +277,7 @@ func (fs *FixationStore) getUnmarshaledEntryForBlock(ctx sdk.Context, safeIndex
 		var entry types.Entry
 		fs.cdc.MustUnmarshal(iterator.Value(), &entry)
 
-		if entry.GetBlock() <= block {
+		if entry.Block <= block {
```
Comment (oren-lava) [inline common/fixation_entry.go:280]: Question: why did you change from ".GetBlock()" to ".Block"? I always use getters to avoid nil dereference. Isn't it a better practice to always use getters?

File: common/fixation_entry.go [hunk 4]
```diff
@@ -308,6 +329,23 @@ func (fs *FixationStore) GetEntry(ctx sdk.Context, index string, entryData codec
...
+// putEntry decrements the refcount of an entry and marks for staleness if needed
+func (fs *FixationStore) putEntry(ctx sdk.Context, entry types.Entry) {
+	if entry.Refcount == 0 {
+		panic("Fixation: prefix " + fs.prefix + ": negative refcount safeIndex: " + entry.Index)
+	}
+
+	entry.Refcount -= 1
+
+	if entry.Refcount == 0 {
+		// never overflows because because ctx.BlockHeight is int64
```
Comment (oren-lava) [inline common/fixation_entry.go:341]: nit: double because

File: common/fixation_entry.go [hunk 5]
```diff
@@ -320,18 +358,12 @@ func (fs *FixationStore) PutEntry(ctx sdk.Context, index string, block uint64) {
...
-	if entry.Refcount == 0 {
-		// never overflows because because ctx.BlockHeight is int64
-		entry.StaleAt = uint64(ctx.BlockHeight()) + uint64(types.STALE_ENTRY_TIME) - 1
+	if entry.Block != block {
+		panic("PutEntry with block mismatch index: " + index +
+			" got " + strconv.Itoa(int(entry.Block)) + " expected " + strconv.Itoa(int(block)))
 	}
 
-	fs.setEntry(ctx, entry)
+	fs.putEntry(ctx, entry)
```
Comment (oren-lava) [inline common/fixation_entry.go:366]: why do we need an additional putEntry? why not put all the code in one function?

**Review by oren-lava — CHANGES_REQUESTED** [review]

File: common/fixation_entry.go [hunk 6]
```diff
@@ -69,6 +71,7 @@ type FixationStore struct {
 	storeKey sdk.StoreKey
 	cdc      codec.BinaryCodec
 	prefix   string
+	tstore   TimerStore
```
Comment (Yaroms) [inline common/fixation_entry.go:74]: lets call it timerStore for clearance

File: common/fixation_entry.go [hunk 7]
```diff
@@ -26,7 +26,9 @@ import (
 //    - GetEntry(index, *entry): get a copy (and reference) of the latest version of an entry
 //    - PutEntry(index, block): drop a reference of a version of an entry
 //    - [TBD] RemoveEntry(index): mark an entry as unavailable for new GetEntry() calls
-//    - GetAllEntryIndex(): get all the entries indices (without versions)
+//    - GetAllEntryIndices(): get all the entries indices (without versions)
```
Comment (Yaroms) [inline common/fixation_entry.go:29]: do we need a module migration here?

File: common/fixation_entry.go [hunk 8]
```diff
@@ -344,8 +382,15 @@ func (fs *FixationStore) createStoreKey(index string) string {
...
+func (fs *FixationStore) AdvanceBlock(ctx sdk.Context) {
+	fs.tstore.Tick(ctx)
+}
+
 // NewFixationStore returns a new FixationStore object
 func NewFixationStore(storeKey sdk.StoreKey, cdc codec.BinaryCodec, prefix string) *FixationStore {
 	fs := FixationStore{storeKey: storeKey, cdc: cdc, prefix: prefix}
+	callback := func(ctx sdk.Context, data string) { fs.deleteStaleEntries(ctx, data) }
+	tstore := NewTimerStore(storeKey, cdc, prefix).WithCallbackByBlockHeight(callback)
+	fs.tstore = *tstore
```
Comment (Yaroms) [inline common/fixation_entry.go:394]: what happens if we set only one type of callback?

**Review by Yaroms — CHANGES_REQUESTED** [review]

File: common/fixation_entry.go (hunk 7 above)
Comment (orenl-lava) [reply common/fixation_entry.go:29]: not needed - just another API that leverages the existing data format.

**Review by Yaroms — APPROVED** [review]

**Review by omerlavanet — APPROVED** [review]
very very nice
